make/photon/prepare/utils/migration.py

```diff
@@ -15,29 +32,35 @@ def read_conf(path):
             exit(-1)
     return d
 
-def _to_module_path(ver):
-    return "migrations.version_{}".format(ver.replace(".","_"))
-
 def search(input_ver: str, target_ver: str) -> deque :
     """
-    Search accept a input version and the target version.
-    Returns the module of migrations in the upgrade path
+    Find the migration path by BFS
+    Args:
+        input_ver(str): The version migration start from
+        target_ver(str): The target version migrated to
+    Returns:
+        deque: the module of migrations in the upgrade path
     """
-    upgrade_path, visited = deque(), set()
-    while True:
-        module_path = _to_module_path(target_ver)
-        visited.add(target_ver)  # mark current version for loop finding
-        if os.path.isdir(os.path.join(MIGRATION_BASE_DIR, 'version_{}'.format(target_ver.replace(".","_")))):
-            module = importlib.import_module(module_path)
-            if module.revision == input_ver:    # migration path found
-                break
-            elif module.down_revision is None:  # migration path not found
-                raise Exception('no migration path found')
-            else:
-                upgrade_path.appendleft(module)
-                target_ver = module.down_revision
-                if target_ver in visited: # version visited before, loop found
-                    raise Exception('find a loop caused by {} on migration path'.format(target_ver))
-        else:
-            raise Exception('{} not dir'.format(os.path.join(MIGRATION_BASE_DIR, 'versions', target_ver.replace(".","_"))))
-    return upgrade_path
+    upgrade_path = deque()
+    parents, visited, q = dict(), set(), deque() # must ensure all version in parents are valid
```

Comment:
The var name parent is a map key is the old version and value is the new version right?

Why call it `parent`?

Reply:
Because the upgrading path is a tree-style graph. When we find the version we expected we need to reverse the finding path. But there is no tree data structure here. So we need a parent array to record the parent.